### DISCUSSION TITLE
crysol for ensemble

### DISCUSSION THREAD

**Discussion by tianyi-shui** [biosaxs-com/atsas-community#1933] [Ab initio shape determination] [answered]
Hi,all
I am new to SAXS data and Just started using crysol software.
I encountered some confusion while using crysol to process ensembles. When I want to obtain the average small-angle scattering profile, should I first fit each conformation to the experimental data, then average the scattering profiles and calculate the chi value, or should I average first and then fit to the experimental data? 
Best wishes
leilei

**Comment by franke-biosaxs** [comment]
Not sure I understand. But before going into more detail, please check out both OLIGOMER and GAJOE, either of the two might be what you are looking for?

**Reply by tianyi-shui** [reply]
Thank you for your response. What I meant to convey is that if I obtain a protein ensemble through simulations and extract thousands of conformations from it, I can use Crysol to predict and calculate the chi value for each conformation. However, I am unsure how to calculate the average scattering profiles of all conformations.

**Reply by franke-biosaxs** [reply]
Ok. Where did you get the ensemble from? If the ensemble is small (up to 50) and you have reason to believe that there is a mixture of the ensemble components in solution, try OLIGOMER. If your ensemble is large (>1000), see GAJOE. Both OLIGOMER and GAJOE use CRYSOL through FFMAKER (`ffmaker --pool` for gajoe). The former is an exact analytic solution to the mixture, it will give you the percentage of each ensemble member in solution, the latter is a numerical approximation attempting to find a small number of ensemble members that represent the data well. Another analytical approach is NNLSJOE (non-negative least squares, almost the same as OLIGOMER, but without limitation of the number of input data sets). This can give you another approximation of the best subset of models that together fit the data.

If you only want to find "the single best fit among all ensemble members", you can also run CRYSOL with all models and the data as input. Be prepared for a wait and a _large_ number of files. You'll need to `grep` the log files for the chi^2.

Does this help?

**Reply by tianyi-shui** [reply]
Thank you very much for your help. My ensemble was obtained through GROMACS simulations and consists of a single component with a large ensemble size (over 1000). In the future, I might assign different weights to various conformations and then calculate their ensemble average to achieve a better chi value. All ensemble members will be used. Can GAJOE handle this? Once again, thank you for your assistance.

**Reply by franke-biosaxs** [reply]
Yes, both GAJOE and NNLSJOE have been implemented with this use case in mind. Again, the difference being: GAjoe is a Genetic Algorithm that returns an approximation that is not unique, i.e. you run GAJOE multiple times, you likely get slightly different results. Contrary, NNLSJOE will pick the single weighted subset across all inputs that, combined, fits the data best. 

IIRC, gromacs is Molecular Dynamics? I.e. based on a starting model you ran time forward and let it "wobble along" (sorry for the term, not my field)? If so, models of subsequent time steps might be too similar to discriminate between them based on noisy SAXS data. If this is the case, maybe only use  every Nth model and drop the intermediate ones to provide larger differences for the programs to work with.

Edit:
Roughly like this:
```
$ ffmaker [prediction options] --output intensities.csv --pool=size.csv
$ gajoe [GA options] --intensity intensities.csv --size size.csv
$ nnlsjoe --intensity intensities.csv --size size.csv --fit nnlsjoe.fit
```
Read the manual and `--help` output to make sure.

**Reply by tianyi-shui** [reply]
Thank you very much for your response. Based on your answer, I tried GAJOE and NNLSJOE. They seem to be selecting and optimizing the input content to obtain the best fitting combination. However, my goal is not optimization, but rather to calculate the average scattering intensity of all input structures and the chi value of the fit after averaging.

**Reply by franke-biosaxs** [reply, marked as answer]
I do not understand why you would want to average the scattering of _everything_, instead of the scattering of only those that "make sense"? You will not obtain a better chi^2 from doing so. 

That said, as you already have the intensities.csv now, average the 2nd to nth column manually (e.g. in Python) and scale the result the experimental data via `datmerge --scale exp.dat average.dat`. Then use `datcmp exp.dat average.dat --test chi-square` to obtain a chi^2.

**Reply by tianyi-shui** [reply]
Thank you very much for your assistance. I aim to use SAXS data as one of the standards for evaluating the quality of the ensemble, so each conformation is extremely important and cannot be discarded. I greatly appreciate your help.

**Reply by tianyi-shui** [reply]
I tested using only one PDB structure. First, I used ffmarker to generate a CSV file, then datmerge --scale exp.dat average.dat to obtain the scaled result. Then, I used datcmp exp.dat average.dat --test chi-square and found that the reported result was completely different from the crysol fit result.
Would it be reasonable to first use crysol to fit each structure to the experimental data, then average the scattering intensities from the .fit files, and finally calculate the chi-square value myself?

**Reply by franke-biosaxs** [reply]
> I tested using only one PDB structure. First, I used ffmarker to generate a CSV file, then datmerge --scale exp.dat average.dat to obtain the scaled result. Then, I used datcmp exp.dat average.dat --test chi-square and found that the reported result was completely different from the crysol fit result. 

crysol adjusts a number of extra parameters during the fitting stage that does not happen in the described procedure. Differences are to be expected.

> Would it be reasonable to first use crysol to fit each structure to the experimental data, then average the scattering intensities from the .fit files, and finally calculate the chi-square value myself?

As each fit would be subject due different parameters (as above), I would say no. You can run crysol on each, then avarage the .int files. But that should effectively be the same to what you described above.

If I were to attempt what you are trying to do, average model scattering with equal weights and fir the average to experimental data, then I'd:
 - run crysol or ffmaker (only difference: the former generates individual files, the latter a single csv file)
 - average the calculated scattering, either with `dataver` or something like Python or Matlab
 - use `datmerge --scale --adjust experimental.dat averaged.dat -o adjusted.dat` to get the scaling and background subtraction (if any)
 - use `datcmp --test chi-square experimental.dat adjusted.dat` to get the chi^2

There is one caveat here in the back of my head: if I remember correctly, the scattering intensity is proportional to the volume squared. Thus you may run into trouble simply averaging everything as-is. Depending on your simulations, you might need to scale by volume first. Think the same sequence of amino acids once fully stretched out, once tightly wound up like a ball. Same weight, completely different volume and almost certainly, completely difference I(0). How you average those, or if you should, is debatable. If all models are "very similar" in size, shape and volume (however that would be defined), then this potential issue can probably be ignored. But in such a case, i.e. your models are essentially identical, you can also simply pick a single model and fit it to your saxs data - given the likely noisiness of said data, SAXS will not be able to differentiate between these similar models and the outcome will be the same.
